### PR TITLE
[SYCL][L0] Restructure code in preparation for submitting to multiple compute queues

### PR DESCRIPTION
**Pull request by smaslov-intel**
Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 1]
```diff
@@ -329,29 +329,41 @@ struct _pi_device : _pi_object {
     // PI device creation.
   }
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.hpp:331]: Please add a small comment about this new data structure.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 2]
```diff
@@ -329,29 +329,41 @@ struct _pi_device : _pi_object {
...
-  int32_t ZeLinkCopyQueueGroupIndex;
-
-  // Keep the index of the compute engine
-  int32_t ZeComputeEngineIndex = 0;
-
-  // Cache the properties of the compute/copy queue groups.
-  ZeStruct<ze_command_queue_group_properties_t> ZeComputeQueueGroupProperties;
-  ZeStruct<ze_command_queue_group_properties_t> ZeMainCopyQueueGroupProperties;
-  ZeStruct<ze_command_queue_group_properties_t> ZeLinkCopyQueueGroupProperties;
+  struct queue_group_info_t {
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.hpp:332]: Please add a small comment about this new data structure.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 3]
```diff
@@ -634,40 +646,49 @@ struct _pi_queue : _pi_object {
...
+    bool isCopy() const { return Type != queue_type::Compute; }
+
+    // Level Zero command queue handles.
+    std::vector<ze_command_queue_handle_t> ZeQueues;
+
+    // This function will return one of possibly multiple available queues.
+    // Currently, a round robin strategy is used.
+    ze_command_queue_handle_t &getZeQueue(uint32_t *QueueGroupOrdinal);
+
+    // These indicies are to filter specific range of the queues to use,
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.hpp:669]: minor typo

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 4]
```diff
@@ -634,40 +646,49 @@ struct _pi_queue : _pi_object {
...
+
+  pi_queue_group_t ComputeQueueGroup{this, queue_type::Compute};
+
   // Vector of Level Zero copy command command queue handles.
-  // Some (or all) of these handles may not be available depending on user
-  // preference and/or target device.
   // In this vector, main copy engine, if available, come first followed by
   // link copy engines, if available.
-  std::vector<ze_command_queue_handle_t> ZeCopyCommandQueues;
+  pi_queue_group_t CopyQueueGroup{this, queue_type::MainCopy};
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.hpp:681]: Please add a small description of the second argument.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 5]
```diff
@@ -406,12 +418,12 @@ struct pi_command_list_info_t {
...
-  // Record the index of copy queue (in the vector of available copy queues)
-  // to which the command list (if any) will be submitted.
-  // If there is no command list, or if the command list is not a copy command
-  // list, the value is set to -1.
-  int CopyQueueIndex{-1};
-  bool isCopy() const { return CopyQueueIndex != -1; }
+  // Record the queue to which the command list will be submitted.
+  ze_command_queue_handle_t ZeQueue{nullptr};
+  // Keeps the ordinal of the ZeQueue queue group. Invalid if ZeQueue==nullptr
+  uint32_t ZeQueueGroupOrdinal{0};
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.hpp:424]: This value is not set anywhere else. Is that an issue?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 6]
```diff
@@ -1025,10 +1028,36 @@ static const zeCommandListBatchConfig ZeCommandListBatchCopyConfig = [] {
...
+        std::min((size_t)FilterUpperIndex, CopyQueues.size() - 1);
+    if (FilterLowerIndex <= FilterUpperIndex) {
+      CopyQueueGroup.ZeQueues = CopyQueues;
+      CopyQueueGroup.LowerIndex = FilterLowerIndex;
+      CopyQueueGroup.UpperIndex = FilterUpperIndex;
+      CopyQueueGroup.NextIndex = CopyQueueGroup.LowerIndex;
+    }
+  }
+
+  // Initialize compute/copy coomand batches.
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1060]: minor typo

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 7]
```diff
@@ -1404,124 +1410,52 @@ bool _pi_queue::isBatchingAllowed(bool IsCopy) const {
...
 // This function also sends back the value of CopyQueueIndex and
 // CopyQueueGroupIndex (optional)
-ze_command_queue_handle_t
-_pi_queue::getZeCopyCommandQueue(int *CopyQueueIndex,
-                                 int *CopyQueueGroupIndex) {
-  assert(CopyQueueIndex);
-  int n = ZeCopyCommandQueues.size();
-  int LowerCopyQueueIndex = getRangeOfAllowedCopyEngines.first;
-  int UpperCopyQueueIndex = getRangeOfAllowedCopyEngines.second;
+ze_command_queue_handle_t &
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1417]: Please modify the comment above

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 8]
```diff
@@ -6047,6 +5992,9 @@ enqueueMemFillHelper(pi_command_type CommandType, pi_queue Queue, void *Ptr,
                      const pi_event *EventWaitList, pi_event *Event) {
   PI_ASSERT(Queue, PI_INVALID_QUEUE);
   PI_ASSERT(Event, PI_INVALID_EVENT);
+  // Pattern size must be a power of two.
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5995]: Does this change belong to this PR?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 9]
```diff
@@ -6056,33 +6004,39 @@ enqueueMemFillHelper(pi_command_type CommandType, pi_queue Queue, void *Ptr,
...
+                .ZeProperties.maxMemoryFillPatternSize < PatternSize) {
+      PreferCopyEngine = false;
+    }
+    if (Device->hasLinkCopyEngine() &&
+        Device->QueueGroup[_pi_device::queue_group_info_t::LinkCopy]
+                .ZeProperties.maxMemoryFillPatternSize < PatternSize) {
+      PreferCopyEngine = false;
+    }
   }
-  // Pattern size must fit the queue.
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.cpp:6078]: Is this check not needed anymore?

**Comment by asudarsa** [comment]
The changes look great. Some minor nitpicks.

**Comment by smaslov-intel** [comment]
Sorry, I had to rebase the changes, but you can see the new comments in this commit: https://github.com/intel/llvm/pull/5617/commits/6718706dd4c30b950212f6afaf4f19644bacc53e

**Review by asudarsa — APPROVED** [review]
LGTM. These modifications look great.

**Review by againull — APPROVED** [review]

**Pull request merged**